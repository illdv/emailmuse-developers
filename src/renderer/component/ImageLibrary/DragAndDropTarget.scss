.dnd-target {
  min-width: 400px;
  min-height: 400px;
  height: 100%;
  width: 100%;
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
}

.dnd-target__overlay {
  background-color: #FFF;
  opacity: 0;
  visibility: hidden;
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  transition: visibility, opacity 0.15s ease;
  outline: 2px dashed #4a4a4a;
  outline-offset: -0.7rem;
  z-index: 1000;
  pointer-events: none;
}

.dnd-target__overlay--can-drop {
  opacity: 0.85 !important;
  visibility: visible !important;
}

.dnd-target__overlay--drag-over {
  background-color: #DDD;
}
