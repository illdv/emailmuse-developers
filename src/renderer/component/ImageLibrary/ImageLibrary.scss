.image-library {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.image-library__container {
  width: 95%;
  height: 95%;
  position: relative;
  display: flex;
  justify-content: space-around;
  flex-wrap: wrap;
}

.image-library__upload-input {
  //padding-top: 1rem;
  position: absolute;
  top: -1000;
  left: -1000;
  opacity: 0;
  visibility: hidden;
}

