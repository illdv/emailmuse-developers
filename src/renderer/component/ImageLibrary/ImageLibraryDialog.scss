.image-library-dialog {
    position: relative;
}

.image-library-dialog__img {
    display: flex;
    margin: auto;
    height: 9rem;
    max-width: 100%;
    overflow: hidden;
    border-radius: 0.25rem;
}

.image-library-dialog__content {
    overflow-y: hidden !important;
}

.image-library-dialog__url-field:hover ~ .image-library-dialog__url-tooltip {
    visibility: visible;
}

.image-library-dialog__url-tooltip {
    visibility: hidden;
    width: 240px;
    background-color: #ececec;
    position: absolute;
    right: 0;
    bottom: -1.9rem;
    text-align: center;
    padding: 5px 0;
}