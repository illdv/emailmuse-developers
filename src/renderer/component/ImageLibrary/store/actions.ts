import * as constants from './constants';
import { createAction } from 'redux-actions';
import { IGetImagesResponse } from 'src/renderer/component/ImageLibrary/store/models';
import * as Status from 'src/renderer/common/PreloaderLayout/Status/constants';

export const getImagesRequest = (page = 1, name: string = '') => ({
  type: constants.GET_IMAGES_REQUEST,
  payload: { page, name},
  meta: { status: Status.LOADING },
});

export const getImagesSuccess = (response: IGetImagesResponse) => ({
  type: constants.GET_IMAGES_SUCCESS,
  payload: response,
  meta: { status: Status.LOADED },
});

export const getImagesFailure = createAction(constants.GET_IMAGES_FAILURE, null, () => ({ status: Status.FAILED }));

export const uploadImagesRequest = (files: File | File[]) => ({
  type: constants.UPLOAD_IMAGES_REQUEST,
  payload: [].concat(files),
  meta: { status: Status.LOADING },
});

export const uploadImagesSuccess =
  createAction(constants.UPLOAD_IMAGES_SUCCESS, null, () => ({ status: Status.LOADED }));

export const uploadImagesFailure =
  createAction(constants.UPLOAD_IMAGES_FAILURE, null, () => ({ status: Status.FAILED }));

export const updateImageRequest = (payload: { imageId: number, name: string }) => ({
  type: constants.UPDATE_IMAGE_REQUEST,
  payload,
  meta: { status: Status.LOADING },
});

export const updateImageSuccess =
  createAction(constants.UPDATE_IMAGE_SUCCESS, null, () => ({ status: Status.LOADED }));

export const updateImageFailure =
  createAction(constants.UPDATE_IMAGE_FAILURE, null, () => ({ status: Status.FAILED }));

export const deleteImagesRequest = (ids: number | number[]) => ({
  type: constants.DELETE_IMAGES_REQUEST,
  payload: [].concat(ids),
  meta: { status: Status.LOADING },
});

export const deleteImagesSuccess =
  createAction(constants.DELETE_IMAGES_SUCCESS, null, () => ({ status: Status.LOADED }));

export const deleteImagesFailure =
  createAction(constants.DELETE_IMAGES_FAILURE, null, () => ({ status: Status.FAILED }));
