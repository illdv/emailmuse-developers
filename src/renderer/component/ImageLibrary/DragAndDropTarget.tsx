import * as React from 'react';
import { DropTarget } from 'react-dnd';
import 'src/renderer/component/ImageLibrary/DragAndDropTarget.scss';
import block from 'bem-ts';
const b = block('dnd-target');

namespace DragAndDropTargetSpace {
  export interface IProps {
    onDrop: (item:any) => void;
    showOverlay?: boolean;
    overlayMessage?: string;
    connectDropTarget?: any;
    isOverCurrent?: boolean;
  }
  export interface IState {

  }
}

const types = {
  FILE: '__NATIVE_FILE__'
};

const specs = {
   drop(props, monitor, component) {
     if (props.onDrop) {
       props.onDrop(monitor.getItem());
     }
  }
};

const collect = (connect, monitor) => ({
  connectDropTarget: connect.dropTarget(),
  isOverCurrent: monitor.isOver({ shallow: true })
});

@DropTarget(types.FILE, specs, collect)
export class DragAndDropTarget extends React.Component<DragAndDropTargetSpace.IProps, DragAndDropTargetSpace.IState> {
  constructor (props) {
    super(props);
  }

  render() {
    const {connectDropTarget, isOverCurrent, showOverlay, overlayMessage, children} = this.props;
    return connectDropTarget(
      <div className={b()}>
        {showOverlay ? <div className={b('overlay-background', {dragging: isOverCurrent})}/> : null}
        {showOverlay && overlayMessage ?
          <div className={b('overlay-message', {dragging: isOverCurrent})}>
            <span>{overlayMessage}</span>
          </div> :
          null
        }
        {children}
      </div>
    );
  }
}
