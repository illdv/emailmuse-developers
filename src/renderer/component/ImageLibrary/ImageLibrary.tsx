import * as React from 'react';
import { IStyle } from 'type/materialUI';
import { Paper, withStyles } from '@material-ui/core/';
import { TypeBackground } from '@material-ui/core/styles/createPalette';
import { DragAndDropTarget } from './DragAndDropTarget';
import { ImageLibraryListComponent } from './ImageLibraryList';
import * as EmailerAPI from 'src/renderer/API/EmailerAPI';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { getImagesRequest } from 'src/renderer/component/ImageLibrary/store/actions';

import 'src/renderer/component/ImageLibrary/ImageLibrary.scss';
import block from 'bem-ts';
const b = block('image-library');

namespace ImageLibrarySpace {
  export interface IProps {
    classes?: any;
    actions?: {
      getImagesRequest: typeof getImagesRequest
    };
  }
  export interface IState {
    items: File[];
  }
}

const styles: IStyle = theme => ({
  root: {
    width: '100%',
    height: '100%',
    backgroundColor: theme.palette.background.paper,
  }
});

const mapStateToProps = state => ({
  // TODO refactor with selectors
  items: state.images.all
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ getImagesRequest }, dispatch)
});

@connect(mapStateToProps, mapDispatchToProps)
class ImageLibrary extends React.Component<ImageLibrarySpace.IProps, ImageLibrarySpace.IState> {
  constructor (props) {
    super(props);
    this.state = {
      items: []
    };
  }

  componentDidMount () {
    this.props.actions.getImagesRequest();
  }

  onDrop = item => {
    if (item && item.files) {
      this.addFiles(item.files);
    }
  }

  onProgress = (percentage) => {
    console.log('Loaded:', percentage);
  }

  // TODO: Should user be prevented from uploading one same image twice
  // and how two images could be identified as 'same'?
  addFiles = files => {
    const filtered = files.filter(file => !this.state.items.some(item => item.name === file.name));
    this.setState({items: [...this.state.items, ...filtered]});
    // tslint:disable-next-line
    // console.log('Files has been added', filtered);

    // const filo = filtered[0];
    // let fileData = new FormData();
    // fileData.append(filo.name, filo.path);

    // EmailerAPI.ImageLibrary.uploadImage(filtered[0], this.onProgress).then(result => {
    //   console.log('Post done', result);
    // });

    // EmailerAPI.ImageLibrary.getImages().then(result => {
    //   console.log('Get done', result);
    // });

    // EmailerAPI.ImageLibrary.updateImage(1, "lol").then(result => {
    //   console.log('Get done', result);
    // });

    // EmailerAPI.ImageLibrary.deleteImages([1,2]).then(result => {
    //   console.log('Get done', result);
    // });
    // filtered[0]
  }

  render() {
    const { classes } = this.props;
    return (
      <Paper elevation={4} className={classes.root}>
        <div className={b()}>
          <DragAndDropTarget
            onDrop={this.onDrop}
            showOverlay={true}
            overlayMessage={'Drop files here to add them to your image library'}
          >
            <div className={b('container')}>
              <ImageLibraryListComponent items={this.state.items}/>
            </div>
          </DragAndDropTarget>
        </div>
      </Paper>
    );
  }
}

export default withStyles(styles)(ImageLibrary as any);
