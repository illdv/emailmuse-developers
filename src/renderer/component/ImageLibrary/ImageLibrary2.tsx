import * as React from 'react';
import { IStyle } from 'type/materialUI';
import { Paper, withStyles } from '@material-ui/core/';
import { TypeBackground } from '@material-ui/core/styles/createPalette';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Button from '@material-ui/core/Button';
import TablePagination from '@material-ui/core/TablePagination';
import block from 'bem-ts';

import {
  deleteImagesRequest,
  getImagesRequest,
  updateImageRequest,
  uploadImagesRequest,
} from 'src/renderer/component/ImageLibrary/store/actions';
import {
  getCurrentPageSelector,
  getImagesURLSelector,
  getLastPageSelector,
  getPerPageSelector,
  getTotalImages,
} from 'src/renderer/component/ImageLibrary/store/selectors';
import { IImageLibraryItem } from 'src/renderer/component/ImageLibrary/store/models';
import 'src/renderer/component/ImageLibrary/ImageLibrary.scss';
import { ImageLibraryDialog } from 'src/renderer/component/ImageLibrary/ImageLibraryDialog';
import { ImageLibraryList } from './ImageLibraryList';
import { DragAndDropTarget } from './DragAndDropTarget';
import { IPagination } from 'src/renderer/common/List/interface';

const b = block('image-library');

namespace ImageLibrarySpace {
  export interface IProps {
    classes?: any;
    actions?: {
      getImagesRequest: typeof getImagesRequest,
      uploadImagesRequest: typeof uploadImagesRequest,
      deleteImagesRequest: typeof deleteImagesRequest,
      updateImageRequest: typeof updateImageRequest,
    };
    items: IImageLibraryItem[];
    pagination: IPagination;
  }

  export interface IState {
    openDialog: boolean;
    chosenImage: IImageLibraryItem;
  }
}

const styles: IStyle = theme => ({
  root: {
    width: '100%',
    height: '100%',
    backgroundColor: theme.palette.background.paper,
  },
});

const mapStateToProps = state => ({
  // items: getImagesURLSelector(state),
  // pagination: {
  //   current_page: getCurrentPageSelector(state),
  //   total: getTotalImages(state),
  //   last_page: getLastPageSelector(state),
  //   per_page: getPerPageSelector(state),
  // },
});

const mapDispatchToProps = dispatch => ({
  // actions: bindActionCreators(
  //   { getImagesRequest, uploadImagesRequest, deleteImagesRequest, updateImageRequest },
  //   dispatch),
});

export class ImageLibrary2 extends React.Component<ImageLibrarySpace.IProps, ImageLibrarySpace.IState> {
  constructor(props) {
    super(props);
    this.state = { openDialog: false, chosenImage: null };
  }

  componentDidMount() {
    // this.props.actions.getImagesRequest();
  }

  onDropFile = item => {
    if (item && item.files) {
      // this.props.actions.uploadImagesRequest(item.files);
    }
  }

  onUploadFiles = e => {
    if (e.target.files) {
      let files = [];
      // tslint:disable-next-line
      for (let i = 0; i < e.target.files.length; i++) {
        files = files.concat(e.target.files[i]);
      }
      if (files.length) {
        // this.props.actions.uploadImagesRequest(files);
      }
    }
  }

  onOpenImageInfo = (item: IImageLibraryItem) => () => {
    this.setState({ openDialog: true, chosenImage: item });
  }

  closeDialog = () => {
    this.setState({ openDialog: false, chosenImage: null });
  }

  deleteItem = (item: IImageLibraryItem) => () => {
    // this.props.actions.deleteImagesRequest(item.id);
  }

  updateItem = (item: IImageLibraryItem, name) => {
    // this.props.actions.updateImageRequest({ imageId: item.id, name });
  }

  // TODO implement onProgress

  onChangePage = (e, page) => {
    // this.props.actions.getImagesRequest(page + 1);
  }

  // TODO: implement properly when there is capability to change rows per currentPage
  onChangeRowsPerPage = e => {
    // console.log('Change rows', e.target.value);
  }

  render() {
    const { classes, pagination } = this.props;
    return (
      <Paper elevation={4} >
        <div className={b()}>
          <TablePagination
            component='div'
            count={pagination.total || 0}
            rowsPerPage={pagination.per_page || 0}
            rowsPerPageOptions={[15]}
            page={pagination.current_page - 1}
            backIconButtonProps={{
              'aria-label': 'Previous Page',
            }}
            nextIconButtonProps={{
              'aria-label': 'Next Page',
            }}
            onChangePage={this.onChangePage}
            onChangeRowsPerPage={this.onChangeRowsPerPage}
          />
          <DragAndDropTarget
            onDrop={this.onDropFile}
            showOverlay={true}
            overlayMessage={'Drop files here to add them to your image library'}
          >
            <div className={b('container')}>
              {
                // this.state.chosenImage &&
                // <ImageLibraryDialog
                //   item={this.state.chosenImage}
                //   onDeleteItem={this.deleteItem}
                //   onUpdateItem={this.updateItem}
                //   onClose={this.closeDialog}
                // />
              }
              <ImageLibraryList
                items={this.props.items}
                onDelete={this.deleteItem}
                onSelect={this.onOpenImageInfo}
              />
            </div>
          </DragAndDropTarget>
          <div className={b('footer')}>
            <Button color='primary'>
              <label htmlFor='upload'>
                Upload images
              </label>
              <input
                id='upload'
                className={b('upload-input')}
                type='file'
                multiple
                onChange={this.onUploadFiles}
              />
            </Button>
          </div>
        </div>
      </Paper>
    );
  }
}

export default withStyles(styles)
(connect(mapStateToProps, mapDispatchToProps)
(ImageLibrary2 as any));
