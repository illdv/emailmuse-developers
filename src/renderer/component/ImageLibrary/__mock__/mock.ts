import { IImageLibraryItem } from '../store/models';

export const mockImage0: IImageLibraryItem = {
  id: 89,
  url:
    'https://infocron.s3.us-east-2.amazonaws.com/user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg.jpeg',
  thumb_url:
    'https://infocron.s3.us-east-2.amazonaws.com/' +
    'user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg_thumb450.jpeg',
  name: 'tesla-cat',
};
export const mockImage1: IImageLibraryItem = {
  id: 90,
  url:
    'https://infocron.s3.us-east-2.amazonaws.com/user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg.jpeg',
  thumb_url:
    'https://infocron.s3.us-east-2.amazonaws.com/' +
    'user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg_thumb450.jpeg',
  name: 'tesla-cat',
};
export const mockImage2: IImageLibraryItem = {
  id: 91,
  url:
    'https://infocron.s3.us-east-2.amazonaws.com/user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg.jpeg',
  thumb_url:
    'https://infocron.s3.us-east-2.amazonaws.com/' +
    'user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg_thumb450.jpeg',
  name: 'tesla-cat',
};
export const mockImageArray: IImageLibraryItem[] = [
  mockImage0,
  mockImage1,
  mockImage2,
];
