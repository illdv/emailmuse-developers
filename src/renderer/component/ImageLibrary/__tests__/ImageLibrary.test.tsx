import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { ImageLibrary as ImageLibrary } from '../ImageLibrary';
import { Provider } from 'react-redux';
import { deleteImagesRequest, getImagesRequest, updateImageRequest, uploadImagesRequest } from '../store/actions';
import { mockImage0, mockImageArray } from '../__mock__/mock';

describe('ImageLibrary', () => {
  describe('>>>initialization', () => {
    let initialState;
    let imageLibraryInitial;
    // let mountedImageLibrary;
    const mockGetImagesRequest = jest.fn();

    beforeEach(() => {
      initialState = {
        items: [],
        pagination: {
          current_page: 1,
          last_page: 3,
          per_page: 15,
          total: 40,
        },
        classes: { root: 'lib' },
        actions: {
          getImagesRequest: mockGetImagesRequest,
          uploadImagesRequest,
          deleteImagesRequest,
          updateImageRequest,
        },
      };
      imageLibraryInitial = shallow(<ImageLibrary {...initialState} />);
      // mountedImageLibrary = mount(<ImageLibrary {...initialState} />);

    });
    // afterEach(() => {
    //   mountedImageLibrary.unmount();
    // });

    it('render ImageLibrary component', () => {
      expect(imageLibraryInitial).toMatchSnapshot();
    });

    // //it doesnt work because context is lost
    // it('should call fetch when mounted', () => {
    //
    //   expect(mountedImageLibrary).toBeDefined();
    //   expect(mockGetImagesRequest).toHaveBeenCalled();
    // });
  });

  describe('ImageLibrary handle event', () => {
    let props; let imageLibrary;
    const mockGetImagesRequest    = jest.fn();
    const mockUploadImagesRequest = jest.fn();
    const mockUpdateImageReturned  = jest.fn();
    const mockDeleteImagesRequest = jest.fn().mockReturnValue(mockUpdateImageReturned);
    const mockUpdateImageRequest  = jest.fn();
    beforeEach(() => {
    props  = {
      items: mockImageArray,
      pagination: {
        current_page: 1,
        last_page: 3,
        per_page: 15,
        total: 39,
      },
      classes: { root: 'lib' },
      actions: {
        getImagesRequest: mockGetImagesRequest,
        uploadImagesRequest: mockUploadImagesRequest,
        deleteImagesRequest: mockDeleteImagesRequest,
        updateImageRequest: mockUpdateImageRequest,
      },
    };
    imageLibrary = shallow(<ImageLibrary {...props} />) as any;
    });

    it('onDeleteItem method from ImageLibraryList or ImageLibraryDialog', () => {
      imageLibrary.instance().deleteItem(mockImage0)();
      expect(mockDeleteImagesRequest).toBeCalledWith(mockImage0.id);
    });

    it('onSelect method from ImageLibraryList ', () => {
      imageLibrary.instance().onOpenImageInfo(mockImage0)();
      expect(imageLibrary.state().openDialog).toBe(true);
      expect(imageLibrary.state().chosenImage).toEqual(mockImage0);
    });

    it('onUpdateItem method from ImageLibraryDialog', () => {
      const testName = 'newNameForImage';
      imageLibrary.instance().updateItem(mockImage0, testName);
      expect(mockUpdateImageRequest).toBeCalledWith({imageId: mockImage0.id, name: testName});
    });

    it('closeDialog method from ImageLibraryDialog', () => {
      imageLibrary.instance().closeDialog();
      expect(imageLibrary.state().openDialog).toBe(false);
      expect(imageLibrary.state().chosenImage).toBe(null);
    });
  });
});
