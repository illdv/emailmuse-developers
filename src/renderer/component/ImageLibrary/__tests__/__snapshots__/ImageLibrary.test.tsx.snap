// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageLibrary >>>initialization render ImageLibrary component 1`] = `
<WithTheme(Fade)
  in={true}
  timeout={1000}
>
  <WithStyles(Paper)
    className="lib"
    elevation={4}
  >
    <div
      className="image-library"
    >
      <Search
        search={[Function]}
      />
      <WithStyles(TablePagination)
        backIconButtonProps={
          Object {
            "aria-label": "Previous Page",
          }
        }
        component="div"
        count={40}
        nextIconButtonProps={
          Object {
            "aria-label": "Next Page",
          }
        }
        onChangePage={[Function]}
        onChangeRowsPerPage={[Function]}
        page={0}
        rowsPerPage={15}
        rowsPerPageOptions={
          Array [
            15,
          ]
        }
      />
      <DropTarget(DragAndDropTarget)
        onDrop={[Function]}
        overlayMessage="Drop files here to add them to your image library"
        showOverlay={true}
      >
        <div
          className="image-library__container"
        >
          <ImageLibraryList
            items={Array []}
            onDelete={[Function]}
            onSelect={[Function]}
          />
        </div>
      </DropTarget(DragAndDropTarget)>
      <div
        className="image-library__footer"
      >
        <WithStyles(Button)
          color="primary"
        >
          <label
            htmlFor="upload"
          >
            Upload images
          </label>
          <input
            className="image-library__upload-input"
            id="upload"
            multiple={true}
            onChange={[Function]}
            type="file"
          />
        </WithStyles(Button)>
      </div>
    </div>
  </WithStyles(Paper)>
</WithTheme(Fade)>
`;
